### PR TITLE
Refactoring/#221 fixed mypy warnings

### LINKED ISSUE #221: Fix mypy warnings
https://blog.whtsky.me/tech/2021/dont-forget-py.typed-for-your-typed-python-package/
https://mypy.readthedocs.io/en/stable/running_mypy.html

### LINKED ISSUE #222: Make BackgroundListenerThread.poller private
https://github.com/exasol/advanced-analytics-framework/blob/fcbf1bf32a3868fe3d85063d4140c31b1f47a623/exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py#L141

### LINKED ISSUE #223: Typo catched -> caught
https://github.com/exasol/advanced-analytics-framework/blob/fcbf1bf32a3868fe3d85063d4140c31b1f47a623/exasol/analytics/query_handler/udf/runner/udf.py#L113C32-L113C39

### PR DESCRIPTION
**Pull request by ckunki**
Closes #221 
Closes #222
Closes #223

### CODE REVIEW

File: exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py [hunk 1]
```diff
@@ -411,18 +434,24 @@ def _create_register_peer_connection(self, message: messages.RegisterPeer):
...
-        self._register_peer_connection = RegisterPeerConnection(
+        return RegisterPeerConnection(
             predecessor=message.source,
             predecessor_send_socket_factory=predecessor_send_socket_factory,
             successor=message.peer,
             successor_send_socket_factory=successor_send_socket_factory,
             my_connection_info=self._my_connection_info,
         )
 
+    @property
```
Comment (ckunki) [inline exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py:445]: Property method `register_peer_connection` checks the private attribute `_register_peer_connection` and raises an exception when it is `None` to avoid code duplication.

Questions: 
* Q1) The names of the private attribute and the property differ on in the prefix `_` and potentially could be mixed up. Should we use different names that can easier be distinguished? 
* Q2) The property is public. Should it be private?

File: exasol/analytics/query_handler/graph/stage/sql/execution/query_handler_state.py [hunk 2]
```diff
@@ -58,20 +59,35 @@ def __init__(
...
+    def current_query_handler_context(self) -> ScopeQueryHandlerContext:
+        value = self._current_query_handler_context
+        if value is None:
+            raise UninitializedAttributeError(
+                "Current query handler context is undefined."
+            )
+        return value
+
+    @property
+    def current_stage(self) -> SQLStage:
```
Comment (ckunki) [inline exasol/analytics/query_handler/graph/stage/sql/execution/query_handler_state.py:80]: Property method `current_stage` checks the private attribute `_current_stage` and raises an exception when it is `None` to avoid code duplication.

Questions: 
* Q1) The names of the private attribute and the property differ on in the prefix `_` and potentially could be mixed up. Should we use different names that can easier be distinguished? 
* Q2) The property is public. Should it be private?

File: exasol/analytics/schema/column_name_builder.py [hunk 3]
```diff
@@ -37,6 +39,8 @@ def with_table_like_name(
         return self
 
     def build(self) -> ColumnName:
+        if self._name is None:
+            raise TypeCheckError("Name must not be None.")
```
Comment (ckunki) [inline exasol/analytics/schema/column_name_builder.py:43]: Is it OK to raise a `TypeCheckError` here (same as typeguard does)?
Raising a different Error such as `ValueError` would require to update the unit tests.

File: exasol/analytics/udf/communication/broadcast_operation.py [hunk 4]
```diff
@@ -54,8 +55,15 @@ def _receive_from_localhost_leader(self) -> bytes:
         self._check_sequence_number(specific_message_obj=specific_message_obj)
         return frames[1].to_bytes()
 
+    @property
```
Comment (ckunki) [inline exasol/analytics/udf/communication/broadcast_operation.py:58]: Property method `multi_node_communicator` checks the private attribute `_multi_node_communicator` and raises an exception when it is `None` to avoid code duplication.

Questions: 
* Q1) The names of the private attribute and the property differ on in the prefix `_` and potentially could be mixed up. Should we use different names that can easier be distinguished? 
* Q2) The property is public. Should it be private?

File: exasol/analytics/udf/communication/gather_operation.py [hunk 5]
```diff
@@ -94,18 +109,25 @@ def _send_local_leader_message_to_multi_node_leader(self):
             local_position=local_position, value_frame=value_frame
         )
 
+    @property
```
Comment (ckunki) [inline exasol/analytics/udf/communication/gather_operation.py:112]: Property method `multi_node_communicator` checks the private attribute `_multi_node_communicator` and raises an exception when it is `None` to avoid code duplication.

Questions: 
* Q1) The names of the private attribute and the property differ on in the prefix `_` and potentially could be mixed up. Should we use different names that can easier be distinguished? 
* Q2) The property is public. Should it be private?

File: exasol/analytics/query_handler/context/proxy/bucketfs_location_proxy.py [hunk 6]
```diff
@@ -1,6 +1,6 @@
 import logging
 
-import exasol.bucketfs as bfs
+import exasol.bucketfs as bfs  # type: ignore[import-untyped]
```
Comment (tkilias) [inline exasol/analytics/query_handler/context/proxy/bucketfs_location_proxy.py:3]: I think, we can do this in the pyproject.toml for all imports, but we probably should add py.typed to bucketfs-python

File: exasol/analytics/query_handler/graph/result.py [hunk 7]
```diff
@@ -79,14 +79,14 @@ def __new__(cls, name: str, bases: Tuple[Type, ...], attrs: Any):
...
-        def _configured_new(cls: Type[cls]):
+        def _configured_new(cls: Type[_T]):
             """This function is called for subclasses of classes that declare _Meta as their metaclass."""
             return _new(cls, result_type)
 
-        result_type.__new__ = _configured_new
-        result_type.__init__ = _init
-        result_type.__setattr__ = _setattr
-        result_type.__delattr__ = _delattr
+        result_type.__new__ = _configured_new  # type: ignore
```
Comment (tkilias) [inline exasol/analytics/query_handler/graph/result.py:86]: yeah, here we probably have no other way.

File: exasol/analytics/query_handler/graph/stage/sql/execution/query_handler_state.py [hunk 8]
```diff
@@ -58,20 +59,35 @@ def __init__(
...
         self,
     ) -> QueryHandler[List[SQLStageInputOutput], SQLStageInputOutput]:
-        self._check_is_valid()
-        return self._current_query_handler
+        value = self._current_query_handler
+        if value is None:
+            raise RuntimeError("No current query handler set.")
+        return value
+
+    @property
```
[the remark this answers is not in the document]
Comment (tkilias) [reply exasol/analytics/query_handler/graph/stage/sql/execution/query_handler_state.py:70]: If the original is private, the new one needs to be private, too. This means it needs another name. Good question, what the name should be. The same is true for current_stage property.  Thinking about a prefix like _checked_*.

File: exasol/analytics/query_handler/udf/runner/udf.py [hunk 9]
```diff
@@ -67,26 +75,38 @@ class QueryHandlerStatus(Enum):
...
+        self._parameter = None
+
+    @property
+    def bucketfs_location(self) -> bfs.path.PathLike:
+        if self._bucketfs_location is None:
+            raise UninitializedAttributeError("BucketFS location is undefined.")
+        return self._bucketfs_location
+
+    @property
+    def parameter(self) -> UDFParameter:
```
Comment (tkilias) [inline exasol/analytics/query_handler/udf/runner/udf.py:102]: these should be private, too

File: exasol/analytics/query_handler/udf/runner/udf.py [hunk 10]
```diff
@@ -67,26 +75,38 @@ class QueryHandlerStatus(Enum):
...
 
     def __init__(self, exa):
         self.exa = exa
-        self.bucketfs_location: Optional[bfs.path.PathLike] = None
-        self.parameter: Optional[UDFParameter] = None
+        self._bucketfs_location = None
+        self._parameter = None
+
+    @property
+    def bucketfs_location(self) -> bfs.path.PathLike:
```
Comment (tkilias) [inline exasol/analytics/query_handler/udf/runner/udf.py:96]: these should be private, too

File: exasol/analytics/query_handler/udf/runner/udf.py [hunk 11]
```diff
@@ -187,10 +211,10 @@ def release_and_create_query_handler_context_if_input_query(
             current_state.top_level_query_handler_context.get_child_query_handler_context()
         )
 
-    def _get_parameter(self, ctx):
+    def _get_parameter(self, ctx) -> UDFParameter:
         iter_num = ctx[0]
```
Comment (tkilias) [inline exasol/analytics/query_handler/udf/runner/udf.py:215]: in another PR, we probably should replace the numbers with named constants

File: exasol/analytics/query_handler/udf/runner/udf.py [hunk 12]
```diff
@@ -260,9 +297,11 @@ def _remove_previous_state(self) -> None:
         self._state_file_bucketfs_location().rm()
 
     def _create_udf_query_result(
-        self, ctx, query_columns: List[Column]
+        self, ctx, query_columns: Optional[List[Column]]
     ) -> UDFQueryResult:
         colum_start_ix = 8 if self.parameter.iter_num == 0 else 4
+        if query_columns is None:
```
Comment (tkilias) [inline exasol/analytics/query_handler/udf/runner/udf.py:303]: questionable, needs more investigation

File: exasol/analytics/query_handler/python_query_handler_runner.py [hunk 13]
```diff
@@ -116,24 +117,28 @@ def _release_and_create_query_handler_context_of_input_query(self):
     def _wrap_return_query(
         self, input_query: SelectQueryWithColumnDefinition
     ) -> Tuple[str, str]:
+        if self._state.input_query_query_handler_context is None:
```
Comment (tkilias) [inline exasol/analytics/query_handler/python_query_handler_runner.py:120]: needs investigation

File: exasol/analytics/schema/column_name_builder.py (hunk 3 above)
Comment (tkilias) [reply exasol/analytics/schema/column_name_builder.py:43]: it is not  a typeguard typecheck, so I would create my own TypeCheckError. However, we can do this also in another PR.

File: exasol/analytics/query_handler/udf/runner/udf.py (hunk 10 above)
Comment (ckunki) [reply exasol/analytics/query_handler/udf/runner/udf.py:96]: OK

File: exasol/analytics/query_handler/udf/runner/udf.py (hunk 9 above)
Comment (ckunki) [reply exasol/analytics/query_handler/udf/runner/udf.py:102]: OK

File: exasol/analytics/udf/communication/broadcast_operation.py (hunk 4 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/broadcast_operation.py]: See last push

File: exasol/analytics/query_handler/graph/stage/sql/execution/query_handler_state.py (hunk 8 above)
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/analytics/query_handler/graph/stage/sql/execution/query_handler_state.py:70]: See last push

File: exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py (hunk 1 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py]: See last push

File: exasol/analytics/udf/communication/gather_operation.py (hunk 5 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/gather_operation.py:112]: See last push

File: exasol/analytics/query_handler/context/proxy/bucketfs_location_proxy.py (hunk 6 above)
Comment (ckunki) [reply exasol/analytics/query_handler/context/proxy/bucketfs_location_proxy.py]: See last push

File: exasol/analytics/query_handler/udf/runner/udf.py (hunk 11 above)
Comment (ckunki) [reply exasol/analytics/query_handler/udf/runner/udf.py:215]: See #228

File: exasol/analytics/query_handler/deployment/lua_script_bundle.py [hunk 14]
```diff
@@ -26,20 +26,19 @@ def __init__(
 
     def bundle_lua_scripts(self, output_buffer: IO):
         with tempfile.TemporaryDirectory() as tmp_dir:
-            tmp_dir = Path(tmp_dir)
-            self.copy_lua_source_files(tmp_dir)
-            self.run_lua_amlg(tmp_dir, output_buffer)
+            path = Path(tmp_dir)
+            self.copy_lua_source_files(path)
+            self.run_lua_amlg(path, output_buffer)
```
Comment (tomuben) [inline exasol/analytics/query_handler/deployment/lua_script_bundle.py:31]:
```suggestion
            tmp_path = Path(tmp_dir)
            self.copy_lua_source_files(tmp_path)
            self.run_lua_amlg(tmp_path, output_buffer)
```

File: exasol/analytics/query_handler/query/result/udf_query_result.py [hunk 15]
```diff
@@ -51,7 +51,7 @@ def rowcount(self) -> int:
 
     def fetch_as_dataframe(
         self, num_rows: Union[str, int], start_col: int = 0
-    ) -> "pandas.DataFrame":
+    ) -> "pandas.DataFrame":  # type: ignore
```
Comment (tomuben) [inline exasol/analytics/query_handler/query/result/udf_query_result.py:54]: Would 
```suggestion
    ) -> Optional[pandas.DataFrame]: 
```
work here?

File: exasol/analytics/query_handler/context/proxy/bucketfs_location_proxy.py [hunk 16]
```diff
@@ -1,4 +1,4 @@
-import logging
+import logging # xtype: ignore[import-untyped]
```
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/analytics/query_handler/context/proxy/bucketfs_location_proxy.py:1]: Sorry, this was a left-overwhile experimenting with mypy settings in `pyproject.toml`.
Will be removed in next push.

File: exasol/analytics/query_handler/query/result/udf_query_result.py (hunk 15 above)
Comment (ckunki) [reply exasol/analytics/query_handler/query/result/udf_query_result.py:54]: Yes, but requires importing pandas. I was not sure whether the string syntax was intentional, here?

File: exasol/analytics/query_handler/udf/runner/udf.py [hunk 17]
```diff
@@ -67,29 +75,41 @@ class QueryHandlerStatus(Enum):
...
 
-@dataclasses.dataclass
+@dataclass
 class UDFResult:
     input_query_view: Optional[str] = None
     input_query: Optional[str] = None
-    final_result = {}
-    query_list = []
-    cleanup_query_list = []
+    final_result: str = "{}"
```
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/analytics/query_handler/udf/runner/udf.py:82]: Good point. Yes, looks like a functional change. On the other hand, after fixing the types a dict did not work anymore, a string did not break the tests and [in one place there is a string assigned](https://github.com/exasol/advanced-analytics-framework/blob/main/exasol/analytics/query_handler/udf/runner/udf.py#L127) to `final_result`, anyway. 

So, in the end, I am not sure about the correct solution.

I was not successful when declaring this attribute as a `Union[str, Dict]`.
Alternatively, we could wrap the stacktrace into a dict?
`final_result = { "stacktrace": stacktrace }`

File: exasol/analytics/query_handler/query/result/udf_query_result.py (hunk 15 above)
Comment (ckunki) [reply exasol/analytics/query_handler/query/result/udf_query_result.py]: Importing pandas can take several seconds. 
See next push, replacing this with an explicit forward reference.

File: exasol/analytics/query_handler/udf/runner/udf.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/analytics/query_handler/udf/runner/udf.py:82]: See also the following code that ensures data type `str` to be used: 

https://github.com/exasol/advanced-analytics-framework/blob/9e3c4da41cb94b8a79679f26926bdb928c780903/exasol/analytics/query_handler/udf/runner/udf.py#L357

File: exasol/analytics/query_handler/query/result/udf_query_result.py (hunk 15 above)
Comment (ckunki) [reply exasol/analytics/query_handler/query/result/udf_query_result.py]: I still get errors. See also https://github.com/pydantic/pydantic/issues/234
As far as I can see, either mypy complains or pytest fails.
I don't see a way to use `Optional` together with a type, that is not imported, yet.

File: exasol/analytics/udf/communication/peer_communicator/background_listener_interface.py [hunk 18]
```diff
@@ -56,79 +76,75 @@ def __init__(
...
+            generic = deserialize_message(received, messages.Message)
+            message = generic.__root__
+            if not isinstance(message, messages.MyConnectionInfo):
+                raise UnexpectedMessageError(
+                    f"Unexpected message of type {type(message)}."
+                )
+            return message.my_connection_info
         except Exception as e:
-            self._logger.exception("Exception", raw_message=message)
+            self._logger.exception("Exception", raw_message=received)
```
Comment (tomuben) [inline exasol/analytics/udf/communication/peer_communicator/background_listener_interface.py:115]: What if the exception happened in self._out_control.socket.receive()?
Then variable `received` is uninitialized, and thus the log would be useless.

File: exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py [hunk 19]
```diff
@@ -96,6 +98,50 @@ def create_background_peer_state_builder() -> BackgroundPeerStateBuilder:
...
+class RuntimeSockets:
+    in_control: Socket
+    out_control: Socket
+    listen: Socket
+    listen_port: int
+    poller: Poller
+
+    @classmethod
+    def create(
+        cls,
```
Comment (tomuben) [inline exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py:111]: Maybe define a generics for `cls`, see https://docs.python.org/3/library/typing.html#generics

File: exasol/analytics/udf/communication/socket_factory/fault_injection.py [hunk 20]
```diff
@@ -64,7 +64,7 @@ def receive(self) -> bytes:
     def receive_multipart(self) -> List[abstract.Frame]:
         message = self._internal_socket.receive_multipart()
         converted_message = [Frame(frame) for frame in message]
-        return converted_message
+        return converted_message  # type: ignore
```
Comment (tomuben) [inline exasol/analytics/udf/communication/socket_factory/fault_injection.py:67]: Why do you need the type ignore comment here?

File: exasol/analytics/udf/communication/socket_factory/zmq_wrapper.py [hunk 21]
```diff
@@ -149,7 +149,7 @@ def poll(
             self._sockets_map[zmq_socket]: _bitmask_to_flags(bitmask)
             for zmq_socket, bitmask in poll_result.items()
         }
-        return result
+        return result  # type: ignore
```
Comment (tomuben) [inline exasol/analytics/udf/communication/socket_factory/zmq_wrapper.py:152]: Why do you need the type ignore here?

File: exasol/analytics/udf/utils/context_wrapper.py [hunk 22]
```diff
@@ -28,7 +25,9 @@ def __getattr__(self, name):
 
     def get_dataframe(
         self, num_rows: Union[str, int], start_col: int = 0
-    ) -> Optional[pd.DataFrame]:
+    ) -> Optional[ForwardRef("pandas.DataFrame")]:  # type: ignore[valid-type]
```
Comment (tomuben) [inline exasol/analytics/udf/utils/context_wrapper.py:28]: Can the "type: ignore" message now be removed?

File: exasol/analytics/udf/utils/iterators.py [hunk 23]
```diff
@@ -1,6 +1,6 @@
 from typing import Any, Callable
 
-import pandas as pd
+import pandas as pd  # type: ignore
```
Comment (tomuben) [inline exasol/analytics/udf/utils/iterators.py:3]: Can be removed?

File: exasol/analytics/utils/errors.py [hunk 24]
```diff
@@ -0,0 +1,12 @@
+class UninitializedAttributeError(Exception):
+    """
+    A method of a class accesses an attribute that has not been
+    initialized dynamically before, e.g. in method run().
+    """
+
+
+class IllegalParametersError(Exception):
+    """
+    Calling a function, method or iniinitializing a class with an illegal
```
Comment (tomuben) [inline exasol/analytics/utils/errors.py:10]:
```suggestion
    Indicating an error  when called a function, method or initialization a class with an illegal
```

**Review by tomuben — CHANGES_REQUESTED** [review]
Don't forget to activate the type checker in the CI !!! Otherwise all the work is in vain ;)

File: exasol/analytics/udf/communication/messages.py [hunk 25]
```diff
@@ -13,7 +13,7 @@ class BaseMessage(BaseModel, frozen=True):
 class RegisterPeer(BaseMessage, frozen=True):
     message_type: Literal["RegisterPeer"] = "RegisterPeer"
     peer: Peer
-    source: Optional["Peer"]
+    source: Optional["Peer"] = None
```
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/analytics/udf/communication/messages.py:16]: Please see the results of my investigation in [discussion_r1860537415](https://github.com/exasol/advanced-analytics-framework/pull/225#discussion_r1860537415) .

TL;DR: not required and not recommended. Using string "Peer" is also shorter.
Only `from __future__ import annotations` could be an option that might allow to drop the quotes.

File: exasol/analytics/udf/communication/peer_communicator/background_listener_interface.py (hunk 18 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/peer_communicator/background_listener_interface.py:115]: The original implementation used variable `message` which is 2 lines further down and hence with even more risk to be uninitialized. 
The logger is a structured logger (I can give additional explanation). 
Hence, a potentially uninitialized variable is acceptable from my perspective. 
We also added a `raise` statement after logging, which was missing before. 

All in all, I think the implementation is significantly improved and hopefully fine. 
But please only agree, if I could convince you.

File: exasol/analytics/udf/communication/socket_factory/fault_injection.py (hunk 20 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/socket_factory/fault_injection.py:67]: I got 
* "List" is invariant -- see https://mypy.readthedocs.io/en/stable/common_issues.html#variance
* Consider using "Sequence" instead, which is covariant
* Perhaps you need a type annotation for "converted_message"? Suggestion: "list[abstract.Frame]"

But adding this type annotation seems to fix the problem.

Comment (ckunki) [reply exasol/analytics/udf/communication/socket_factory/fault_injection.py:67]: See next push.

File: exasol/analytics/udf/communication/socket_factory/zmq_wrapper.py (hunk 21 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/socket_factory/zmq_wrapper.py:152]: indeed - not needed. Thanks. 
See next push removing the ignore comment.

File: exasol/analytics/udf/utils/context_wrapper.py (hunk 22 above)
Comment (ckunki) [reply exasol/analytics/udf/utils/context_wrapper.py:28]: Currently, I don't see any perfect solution and hence propose to keep the comment.

Please see the results of my investigation in [discussion_r1860537415](https://github.com/exasol/advanced-analytics-framework/pull/225#discussion_r1860537415) .

**Comment by ckunki** [comment]
> Don't forget to activate the type checker in the CI !!! Otherwise all the work is in vain ;)

I assume the type checker was activated all the time in [checks.yml](https://github.com/exasol/advanced-analytics-framework/blob/main/.github/workflows/checks.yml#L62):        
```
run: poetry run nox -s type-check
```

but mypy was told to ignore all package of the project in [pyproject.toml](https://github.com/exasol/advanced-analytics-framework/blob/main/pyproject.toml#60)

File: exasol/analytics/query_handler/udf/runner/udf.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (ckunki) [reply exasol/analytics/query_handler/udf/runner/udf.py:82]: See #229

File: exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py (hunk 19 above)
Comment (ckunki) [reply exasol/analytics/udf/communication/peer_communicator/background_listener_thread.py:111]: This is a class method, compare to `self`.

**Review by tomuben — APPROVED** [review]

File: exasol/analytics/query_handler/udf/runner/udf.py (hunk 17 above)
[the remark this answers is not in the document]
Comment (tkilias) [reply exasol/analytics/query_handler/udf/runner/udf.py:82]: Fixing of the missing Generic Types for QueryHandlerRunnerState will happen in https://github.com/exasol/advanced-analytics-framework/issues/230

**Review by tkilias — APPROVED** [review]

File: pyproject.toml [hunk 26]
```diff
@@ -26,9 +26,7 @@ importlib-resources = "^6.4.0"
 exasol-bucketfs = ">=0.6.0,<1.0.0"
 click = "^8.0.4"
 pyexasol = ">=0.25.0,<1.0.0"
-pandas = "^1.1.0"
-# request from Mikhail:
-# pandas = "^2.2.3"
+pandas = "^2.2.3"
 pydantic = "^1.10.2"
```
Comment (tkilias) [inline pyproject.toml:30]: We should upgrade pydantic to version 2 it might otherwise conflict with other projects

Comment (ckunki) [reply pyproject.toml:30]: See ticket
* #233
